File: packages/db/src/enums.ts
// Single source of truth for column enums shared across layers. The value
// tuples live here; `schema.ts` feeds them to Drizzle's `text({ enum })`,
// `schemas.ts` (shared) feeds them to `z.enum`, and `responses.ts` derives its
// literal-union types from them — so DB, API input, and API output can never
// drift out of sync.

export const MAILBOX_TYPES = ["personal", "group", "service", "temp"] as const;
export type MailboxType = (typeof MAILBOX_TYPES)[number];

export const MESSAGE_DIRECTIONS = ["in", "out"] as const;
export type MessageDirection = (typeof MESSAGE_DIRECTIONS)[number];

export const USER_ROLES = ["admin", "user"] as const;
export type UserRole = (typeof USER_ROLES)[number];

export const SPAM_FILTER_LEVELS = ["off", "auth", "standard", "ai"] as const;
export type SpamFilterLevel = (typeof SPAM_FILTER_LEVELS)[number];

export const SERVICE_MODES = ["duplex", "send"] as const;
export type ServiceMode = (typeof SERVICE_MODES)[number];

export const SPAM_VERDICTS = ["clean", "suspicious", "spam"] as const;
export type SpamVerdict = (typeof SPAM_VERDICTS)[number];

// AI auto-category assigned to inbound mail (best-effort, Workers AI). A fixed
// taxonomy; each tag is defined for the model in `mail/ai.ts` (CATEGORY_GUIDE)
// and coloured as a chip in the list (thread-row.tsx). Keep the three in sync.
// `other` is the fallback when nothing fits.
export const AI_CATEGORIES = [
  "personal",
  "newsletter",
  "promotion",
  "shipping",
  "receipt",
  "finance",
  "travel",
  "social",
  "security",
  "update",
  "notification",
  "calendar",
  "other",
] as const;
export type AiCategory = (typeof AI_CATEGORIES)[number];

// AI-judged importance of inbound mail (best-effort, Workers AI). Drives which
// kind of push notification a mailbox sends. "high" only for genuinely
// time-sensitive/personal mail; "low" for bulk/automated.
export const AI_PRIORITIES = ["high", "normal", "low"] as const;
export type AiPriority = (typeof AI_PRIORITIES)[number];

// Per-mailbox, per-priority-tier notification style. "none" = silent,
// "important" = attention-grabbing (sticky/vibrate).
export const NOTIFY_LEVELS = ["none", "normal", "important"] as const;
export type NotifyLevel = (typeof NOTIFY_LEVELS)[number];

export const EDITOR_FORMATS = ["text", "markdown", "html"] as const;
export type EditorFormat = (typeof EDITOR_FORMATS)[number];

export const QUOTE_KINDS = ["reply", "forward"] as const;
export type QuoteKind = (typeof QUOTE_KINDS)[number];

// Per-mailbox PGP policy: off | sign outbound | sign + encrypt outbound.
export const PGP_MODES = ["off", "sign", "sign_encrypt"] as const;
export type PgpMode = (typeof PGP_MODES)[number];

// Inbound signature verification outcome stored on a message.
export const PGP_VERIFY = ["good", "bad", "unknown"] as const;
export type PgpVerify = (typeof PGP_VERIFY)[number];

// How a stored correspondent public key was obtained. `tofu` = captured from an
// inbound signed/attached key; `wkd` = auto-fetched from the sender's Web Key
// Directory; `import` = added by the owner.
export const CONTACT_KEY_SOURCES = ["import", "tofu", "wkd"] as const;
export type ContactKeySource = (typeof CONTACT_KEY_SOURCES)[number];

// A per-message PGP key event the reader surfaces as a banner: a contact key was
// auto-captured (TOFU/WKD) for the first time, or a known contact signed with a
// different key than the one on file (possible rotation — or impersonation).
export const PGP_KEY_EVENTS = ["captured", "rotated"] as const;
export type PgpKeyEvent = (typeof PGP_KEY_EVENTS)[number];

// A manual blocklist entry targets either a single address or a whole domain.
export const BLOCK_ENTRY_TYPES = ["email", "domain"] as const;
export type BlockEntryType = (typeof BLOCK_ENTRY_TYPES)[number];

// Lifecycle of a user-submitted block request awaiting admin review.
export const BLOCK_REQUEST_STATUS = ["pending", "approved", "denied"] as const;
export type BlockRequestStatus = (typeof BLOCK_REQUEST_STATUS)[number];

// ─── Reminders ──────────────────────────────────────────────────────────────

// A reminder is either set manually on a thread ("remind me about this") or
// created automatically when a mail is sent ("remind me if no reply").
export const REMINDER_KINDS = ["manual", "follow_up"] as const;
export type ReminderKind = (typeof REMINDER_KINDS)[number];

// Lifecycle: pending until its time arrives → fired (surfaced in the bell) →
// done (user dismissed it). A follow-up flips to cancelled when a reply lands.
export const REMINDER_STATUSES = ["pending", "fired", "done", "cancelled"] as const;
export type ReminderStatus = (typeof REMINDER_STATUSES)[number];

// ─── Inbound rules / filters ────────────────────────────────────────────────

// Message field a rule condition matches against. `deliveredTo` is the envelope
// recipient (the specific incoming address, incl. plus-aliases).
export const RULE_FIELDS = ["from", "to", "cc", "subject", "body", "deliveredTo"] as const;
export type RuleField = (typeof RULE_FIELDS)[number];

// How a condition value is compared. `wildcard` supports `*`/`?` globs; `regex`
// is opt-in power-user matching (compiled with safety caps, see mail/rules.ts).
export const RULE_OPS = [
  "contains",
  "equals",
  "startsWith",
  "endsWith",
  "wildcard",
  "regex",
] as const;
export type RuleOp = (typeof RULE_OPS)[number];

// Combinator across a rule's conditions: all = AND, any = OR.
export const RULE_CONDITION_MODES = ["all", "any"] as const;
export type RuleConditionMode = (typeof RULE_CONDITION_MODES)[number];

// What a matched rule does. `stopProcessing` halts evaluation of lower-priority
// rules; `hardBlock` SMTP-rejects the message (nothing stored). `forward` and
// `autoReply` are best-effort outbound sends fired after the message is stored —
// they never block delivery (see mail/rule-sends.ts).
export const RULE_ACTION_TYPES = [
  "applyLabel",
  "moveFolder",
  "markRead",
  "markSpam",
  "forward",
  "autoReply",
  "hardBlock",
  "stopProcessing",
] as const;
export type RuleActionType = (typeof RULE_ACTION_TYPES)[number];

File: apps/ios/CFMail/Core/Models/Enums.swift
import Foundation

// Mirrors packages/db/src/enums.ts. Enums the server owns decode leniently
// where a new variant is plausible (AI taxonomy), strictly where the tuple is
// a closed contract (direction, flags, folder kinds).

nonisolated enum MailboxType: String, Codable, Sendable, CaseIterable {
    case personal, group, service, temp
}

nonisolated enum MessageDirection: String, Codable, Sendable {
    case incoming = "in"
    case outgoing = "out"
}

nonisolated enum UserRole: String, Codable, Sendable {
    case admin, user
}

nonisolated enum SpamFilterLevel: String, Codable, Sendable, CaseIterable, Identifiable {
    case off, auth, standard, ai
    var id: String { rawValue }

    var title: String {
        switch self {
        case .off: "Off"
        case .auth: "Authentication only"
        case .standard: "Standard"
        case .ai: "Standard + AI"
        }
    }

    var detail: String {
        switch self {
        case .off: "Nothing is scored. Everything lands in the inbox."
        case .auth: "Only SPF/DKIM/DMARC failures count against a message."
        case .standard: "Authentication, reputation and content heuristics."
        case .ai: "Adds a Workers AI pass for messages the heuristics can't call."
        }
    }
}

nonisolated enum SpamVerdict: String, Codable, Sendable {
    case clean, suspicious, spam
}

nonisolated enum AiCategory: String, Codable, Sendable, CaseIterable {
    case personal, newsletter, promotion, shipping, receipt, finance
    case travel, social, security, update, notification, calendar, other

    nonisolated init(from decoder: any Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = AiCategory(rawValue: raw) ?? .other
    }
}

nonisolated enum AiPriority: String, Codable, Sendable, CaseIterable {
    case high, normal, low

    nonisolated init(from decoder: any Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = AiPriority(rawValue: raw) ?? .normal
    }
}

nonisolated enum NotifyLevel: String, Codable, Sendable, CaseIterable, Identifiable {
    case none, normal, important
    var id: String { rawValue }

    var title: String {
        switch self {
        case .none: "Silent"
        case .normal: "Normal"
        case .important: "Time sensitive"
        }
    }
}

nonisolated enum EditorFormat: String, Codable, Sendable, CaseIterable, Identifiable {
    case text, markdown, html
    var id: String { rawValue }

    var title: String {
        switch self {
        case .text: "Plain text"
        case .markdown: "Markdown"
        case .html: "HTML"
        }
    }
}

nonisolated enum QuoteKind: String, Codable, Sendable {
    case reply, forward
}

nonisolated enum PgpMode: String, Codable, Sendable, CaseIterable, Identifiable {
    case off
    case sign
    case signEncrypt = "sign_encrypt"
    var id: String { rawValue }

    var title: String {
        switch self {
        case .off: "Off"
        case .sign: "Sign outbound"
        case .signEncrypt: "Sign + encrypt outbound"
        }
    }
}

nonisolated enum PgpVerify: String, Codable, Sendable {
    case good, bad, unknown
}

nonisolated enum ContactKeySource: String, Codable, Sendable {
    case importedKey = "import"
    case tofu
    case wkd

    var label: String {
        switch self {
        case .importedKey: "Imported"
        case .tofu: "Captured from mail"
        case .wkd: "Web Key Directory"
        }
    }
}

nonisolated enum PgpKeyEvent: String, Codable, Sendable {
    case captured, rotated
}

nonisolated enum ReminderKind: String, Codable, Sendable {
    case manual
    case followUp = "follow_up"
}

nonisolated enum ReminderStatus: String, Codable, Sendable {
    case pending, fired, done, cancelled
}

nonisolated enum RuleField: String, Codable, Sendable, CaseIterable, Identifiable {
    case from, to, cc, subject, body
    case deliveredTo
    var id: String { rawValue }

    var title: String {
        switch self {
        case .from: "From"
        case .to: "To"
        case .cc: "Cc"
        case .subject: "Subject"
        case .body: "Body"
        case .deliveredTo: "Delivered to"
        }
    }
}

nonisolated enum RuleOp: String, Codable, Sendable, CaseIterable, Identifiable {
    case contains, equals, startsWith, endsWith, wildcard, regex
    var id: String { rawValue }

    var title: String {
        switch self {
        case .contains: "contains"
        case .equals: "is exactly"
        case .startsWith: "starts with"
        case .endsWith: "ends with"
        case .wildcard: "matches (glob)"
        case .regex: "matches (regex)"
        }
    }
}

nonisolated enum RuleConditionMode: String, Codable, Sendable, CaseIterable, Identifiable {
    case all, any
    var id: String { rawValue }
    var title: String { self == .all ? "Match all conditions" : "Match any condition" }
}

nonisolated enum BlockEntryType: String, Codable, Sendable {
    case email, domain
}

/// The list views a mailbox exposes. Mirrors `MAIL_VIEWS` in shared/schemas.ts.
nonisolated enum MailView: String, Codable, Sendable, CaseIterable, Identifiable {
    case inbox, drafts, sent, marked, spam, trash, all
    var id: String { rawValue }

    var title: String {
        switch self {
        case .inbox: "Inbox"
        case .drafts: "Drafts"
        case .sent: "Sent"
        case .marked: "Starred"
        case .spam: "Spam"
        case .trash: "Trash"
        case .all: "All Mail"
        }
    }

    var symbol: String {
        switch self {
        case .inbox: "tray"
        case .drafts: "doc"
        case .sent: "paperplane"
        case .marked: "star"
        case .spam: "xmark.bin"
        case .trash: "trash"
        case .all: "tray.full"
        }
    }

    /// Mail tints every mailbox glyph with the app colour except Flagged.
    var tintIsOrange: Bool { self == .marked }

    /// Views whose badge counts unread rather than total.
    var badgeCountsUnread: Bool { self == .inbox || self == .spam }
}

nonisolated enum SearchIn: String, Codable, Sendable, CaseIterable, Identifiable {
    case all, subject, from, body
    var id: String { rawValue }
    var title: String { self == .all ? "Everywhere" : rawValue.capitalized }
}

nonisolated enum SearchFolder: String, Codable, Sendable, CaseIterable, Identifiable {
    case any, inbox, sent, marked, spam, trash
    var id: String { rawValue }

    var title: String {
        switch self {
        case .any: "Anywhere"
        case .marked: "Starred"
        default: rawValue.capitalized
        }
    }
}

nonisolated enum SearchSort: String, Codable, Sendable, CaseIterable, Identifiable {
    case newest, oldest, relevance
    var id: String { rawValue }
    var title: String { rawValue.capitalized }
}

// ─── Bit flags ──────────────────────────────────────────────────────────────

/// Message flag bits (shared/flags.ts). `DELETED` is IMAP-only and never shown.
nonisolated enum Flag {
    static let seen = 1 << 0
    static let starred = 1 << 1
    static let draft = 1 << 2
    static let sent = 1 << 3
    static let trash = 1 << 4
    static let deleted = 1 << 5
}

/// Per-mailbox permission bits (shared/permissions.ts).
nonisolated enum Perm {
    static let read = 1 << 0
    static let write = 1 << 1
    static let manage = 1 << 2
    static let all = read | write | manage
}

nonisolated func hasBit(_ value: Int, _ bit: Int) -> Bool { value & bit == bit }
